File: uva_558.py
num=int(input())
while num:
    num-=1
    n, m=input().split()
    n, m=int(n), int(m)
    v=[]
    dis=[1]
    for i in range(m):
        data=input().split()
        temp=[]
        for item in data:
            temp.append(int(item))
        v.append(temp)

    for i in range(1, n):
        dis.append(1000000000)
    for i in range(n):
        for j in range(m):
            dis[v[j][1]]=min(dis[v[j][1]], dis[v[j][0]]+v[j][2])
    check=True
    for i in range(m):
        if dis[v[i][1]]>dis[v[i][0]]+v[i][2]:
            print('possible')
            check=False
            break
    if check:
        print('not possible')

File: uva_558.cpp
# include <vector>
# include <iostream>

using namespace std;
int main(void){
    int num;
    vector<int*> v;
    cin >> num;
    while(num--){
        int n, m;
        while(cin >> n >> m){
            v.clear();
            vector<int> dis(1, 0);
            for (int i=0; i<m; ++i){
                int *data=new int[3];
                cin >> data[0]>>data[1]>>data[2];
                v.push_back(data);
            }
            for (int i=1; i<n; ++i)
                dis.push_back(1000000000);

            for (int i=0; i<n; ++i)
                for (int j=0; j<m; ++j)
                    dis[v[j][1]]=min( dis[v[j][1]], dis[v[j][0]]+v[j][2] );

            bool check=true;
            for (int i=0; i<m ; ++i)
                if( dis[v[i][1]] > dis[v[i][0]]+v[i][2]) {
                    check=false;
                    cout << "possible\n";
                    break;
                }
            if (check)
                cout << "not possible\n";
        }
    }
    return 0;
}